### PR TITLE
HHH-13130 - Provide Gradle-based bytecode enhancement as a task separ…

### PR DESCRIPTION
**Pull request by sebersole**
…ate from the compileJava task

### CODE REVIEW

**Comment by sebersole** [comment]
Cherry-pick from the commit in master

File: tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+		List<URL> urls = new ArrayList<>();
+		for ( File file : runtimeClasspath ) {
+			try {
+				urls.add( file.toURI().toURL() );
+			}
+			catch (MalformedURLException e) {
+				throw new GradleException( "Unable to resolve classpath entry to URL : " + file.getAbsolutePath(), e );
+			}
+		}
+		return new URLClassLoader( urls.toArray( new URL[0] ), Enhancer.class.getClassLoader() );
```
Comment (gbadner) [inline tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java]: The old line in ```HibernatePlugin``` was:

```		return new URLClassLoader( urls.toArray( new URL[urls.size()] ), Enhancer.class.getClassLoader() );```

Was  ```new URL[urls.size()]``` supposed to be changed to ```urls.toArray( new URL[0] )```?

**Comment by gbadner** [comment]
@sebersole, I'm not fluent in Gradle. The [documentation](https://docs.jboss.org/hibernate/orm/5.3/userguide/html_single/Hibernate_User_Guide.html#BytecodeEnhancement-enhancement-gradle) says that the Gradle plugin can be used for build-time enhancement.

I'm not fluent in Gradlew. Does this change make it possible to enhance classes that are already built?

If so, the documentation should be updated.

File: tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java (hunk 1 above)
Comment (sebersole) [reply tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java]: Building an array of size 0 is the new recommended way to do this.  The JVM can already dynamically resize an array very efficiently.

Comment (gbadner) [reply tooling/hibernate-gradle-plugin/src/main/groovy/org/hibernate/orm/tooling/gradle/EnhancementHelper.java]: OK, good to know.

**Comment by sebersole** [comment]
Thank you IntelliJ :)

**Comment by gbadner** [comment]
> > I'm not fluent in Gradlew. 

The above was an edit mistake. You must have copied it before I changed it.

I am trying to understand the use case so that I can try it out. I'm discussing with Stephen...

**Pull request merged**